multiversx/mx-chain-go
# Feat/transaction ordering on execution

## Reasoning behind the pull request
The grouping of transactions in the block give precedence to source and destination shards (transactions are grouped into miniblocks, by the sender and destination shards, each miniblock having only one shard as source for the transactions inside (sender shard) and one shard as destination for the transactions inside(destination shard)). This is useful for the cross shard interactions to be optimized and correctly tracked by the metachain nodes, but th

Currently the transactions execution order is re-computed after the actual execution and fed into the outport driver, which serves clients such as indexers or notifiers. In some cases, e.g multiple smart contract results generated by the same transaction, that need to be executed cross shard in the same destination shard, the execution order for these individual SCRs cannot be correctly determined after execution and all will be treated as a batch and receive the same order.

The execution order of SCRs is currently estimated (post processing) rather than taken from execution, which makes it less efficient and in some cases is more prone to errors. The SCRs ordering will also become available from the execution component directly, together with the execution order of the other transactions and integrate the ordering into the outport driver.
  
## Proposed changes
Create an ordered collection component that can be used during the Txs execution and collect the transactions and scrs in their correct execution order.This feature introduces this component and its integration.

On generation (execution output results), the smart contract results should be added to the ordering component, which should be made available on the post processing component as well as on the processors and preprocessors (already done in the feat branch).

The ordering component will give the correct relative order to the generated SCRs and transactions upon insertion.
If the SCRs are intra shard, then the given order means generation order as well as execution order.

The outport data provider needs to updated with this component and take the ordering from there instead of running the re-estimation.

## Testing procedure
- Created unit tests that fully cover the component.
- Regular system test for non regressions. 
- Run the reconciliation scripts that compare the execution logs with the order given by the component, for each of the committed blocks
- Check indexed blocks, for every type of transaction - regular move balance tx, relayed, builtin function call, rewards tx, smart contract result, contract call, for both successful and failed txs and check that the ordering is correct.

## Pre-requisites
- adapt the execution order validation scripts
- enable the debug component for the execution order during testing 

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?
